pkg/cli/cli.go

```diff
@@ -248,20 +335,90 @@ func (c cli) buildRootCmd() *cobra.Command {
 	return rootCmd
 }
 
+// filterPluginsByKeys matches plugins known to a cli to a set of plugin keys,
+// returning unambiguously matched plugins.
+// A match occurs if:
+// - Name and version are the same.
+// - Long or short name is the same, and only one version for that name exists.
+// If long or short name is the same, but multiple versions for that name exist,
+// don't guess which version to use and instead return an error.
+func filterPluginsByKeys(versionedPlugins []plugin.Base, cliPluginKeys map[string]string) ([]plugin.Base, error) {
+	if len(cliPluginKeys) == 0 {
+		return versionedPlugins, nil
+	}
```

Comment:
I think this depends on the semantics of your filter. If a filter is empty then all plugins pass through it correct? Anyway this function is used to determine if the values passed to `--plugins` contain ambiguous names, so if there are no names to filter by then we can rely on the binding functions for `Init`, `CreateAPI`, and `CreateWebhook` to choose the right plugin to run. Perhaps the naming isn't great; `findRelevantPlugins` instead?

Reply:
If I understood this right, `cliPluginVersion` is a whitelist not a blacklist, so I expect an empty whitelist to yield an empty result, while if it was an empty blacklist I would expect a full result. Maybe I'm not interpreting it the same way you do.

Reply:
To understand the filtering semantics I went back to the `package main` entrypoint:

<details><summary>Step by step walk to know which are the input params</summary>

Entrypoint: [`func main() {...}`](https://github.com/estroz/kubebuilder/blob/2003c53102f972a8b3b65ca3cb38014dd3dc6d36/cmd/main.go#L28-L46)
It does two things: `cli.New(...)` and `c.Run()` with their error handlers. [`c.Run()`](https://github.com/estroz/kubebuilder/blob/2003c53102f972a8b3b65ca3cb38014dd3dc6d36/pkg/cli/cli.go#L101-L104) is only calling `Execute` on the `*cobra.Command` so let's focus on [`cli.New(...)`](https://github.com/estroz/kubebuilder/blob/2003c53102f972a8b3b65ca3cb38014dd3dc6d36/pkg/cli/cli.go#L81-L99).
First it sets some values for the struct and allocates some maps. Then it parses the provided options which set additional fields (you can find the options in [`pkg/cli/cli.go:106-145`](https://github.com/estroz/kubebuilder/blob/2003c53102f972a8b3b65ca3cb38014dd3dc6d36/pkg/cli/cli.go#L106-L145)). And when the fields are set, it calls [`c.initialize()`](https://github.com/estroz/kubebuilder/blob/2003c53102f972a8b3b65ca3cb38014dd3dc6d36/pkg/cli/cli.go#L147-L208).
The first thing `initialize` does is to parse some common flags in [`c.parseBaseFlags()`](https://github.com/estroz/kubebuilder/blob/2003c53102f972a8b3b65ca3cb38014dd3dc6d36/pkg/cli/cli.go#L210-L245). This is important as [one of them is the list of plugins](https://github.com/estroz/kubebuilder/blob/2003c53102f972a8b3b65ca3cb38014dd3dc6d36/pkg/cli/cli.go#L223) which are later [parsed and stored](https://github.com/estroz/kubebuilder/blob/2003c53102f972a8b3b65ca3cb38014dd3dc6d36/pkg/cli/cli.go#L231-L242) in `c.cliPluginKeys`. This parsing does a first duplicate check, it doesn't accept multiple plugins with the same name, for example `go/v1,go/v2` or `go/v2,go/v2` will yield errors.
Then `initialize` reads the configuration file and validates the already set fields, but none of these modify them. So [by this point](https://github.com/estroz/kubebuilder/blob/2003c53102f972a8b3b65ca3cb38014dd3dc6d36/pkg/cli/cli.go#L175) we have in `c.cliPluginKeys` a map of plugin names to versions as passed in the command flag without duplicates and with valid names and versions. This is one of the parameters to the function we are studying. The other one is obtained with [`c.getVersionedPlugins()`](https://github.com/estroz/kubebuilder/blob/2003c53102f972a8b3b65ca3cb38014dd3dc6d36/pkg/cli/cli.go#L408-L419) which basically returns the slice of plugins configured for the current scaffold version with [`WithPlugins`](https://github.com/estroz/kubebuilder/blob/2003c53102f972a8b3b65ca3cb38014dd3dc6d36/pkg/cli/cli.go#L123-L136) option to `cli.New`.
So the input to our function are a list of the plugins that kubebuilder has and the list of plugins requested in the command line.
</details>

First of all, in case a configuration file is read successfully, I think we need to add the plugins found under `layout` to the requested ones in the command line, right? It is only reading the project version: https://github.com/estroz/kubebuilder/blob/2003c53102f972a8b3b65ca3cb38014dd3dc6d36/pkg/cli/cli.go#L163-L166.

And after that, if we don't get any plugins requested (`len(cliPluginKeys) == 0`), I think we should use the latest `go` plugin for that scaffold version (we may need a global map to track this).

Reply:
The goal of `filterPluginsByKeys` is to return a basic set of plugins if `--plugins` is not set, which it currently accomplishes but fails to consider a few cases as you mention. It also breaks semantics of a whitelist, which is what `cliPluginKeys` should be.

A new option `WithDefaultPlugins` that sets a field `defaultPluginsFromOptions` containing default plugins to run if `--plugins` is unset should be added since returning a default list of plugins if the whitelist is empty makes sense.

Unknown plugins cause an error to be returned:
```console
$ kubebuilder init --plugins foo
2020/03/05 10:44:00 no plugins found for CLI plugin keys ["foo"], likely a naming discrepancy
```